activation/certifier.go

```diff
@@ -0,0 +1,369 @@
+package activation
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+	"github.com/spacemeshos/poet/shared"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/sql"
+	"github.com/spacemeshos/go-spacemesh/sql/localsql"
+	"github.com/spacemeshos/go-spacemesh/sql/localsql/certifier"
+)
+
+type CertifierClientConfig struct {
+	// Base delay between retries, scaled with the number of retries.
+	RetryDelay time.Duration `mapstructure:"retry-delay"`
+	// Maximum time to wait between retries
+	MaxRetryDelay time.Duration `mapstructure:"max-retry-delay"`
+	// Maximum number of retries
+	MaxRetries int `mapstructure:"max-retries"`
+}
+
+type Base64Enc struct {
+	Inner []byte
+}
+
+func (e Base64Enc) String() string {
+	return base64.RawStdEncoding.EncodeToString(e.Inner)
+}
+
+// Set implements pflag.Value.Set.
+func (e *Base64Enc) Set(value string) error {
+	return e.UnmarshalText([]byte(value))
+}
+
+// Type implements pflag.Value.Type.
+func (Base64Enc) Type() string {
+	return "Base64Enc"
+}
+
+func (e *Base64Enc) UnmarshalText(text []byte) error {
+	b, err := base64.StdEncoding.DecodeString(string(text))
+	if err != nil {
+		return err
+	}
+	e.Inner = b
+	return nil
+}
+
+type CertifierConfig struct {
+	Client CertifierClientConfig `mapstructure:"client"`
+}
+
+func DefaultCertifierClientConfig() CertifierClientConfig {
+	return CertifierClientConfig{
+		RetryDelay:    1 * time.Second,
+		MaxRetryDelay: 30 * time.Second,
+		MaxRetries:    5,
+	}
+}
+
+func DefaultCertifierConfig() CertifierConfig {
+	return CertifierConfig{
+		Client: DefaultCertifierClientConfig(),
+	}
+}
+
+type ProofToCertify struct {
+	Nonce   uint32 `json:"nonce"`
+	Indices []byte `json:"indices"`
+	Pow     uint64 `json:"pow"`
+}
+
+type ProofToCertifyMetadata struct {
+	NodeId          []byte `json:"node_id"`
+	CommitmentAtxId []byte `json:"commitment_atx_id"`
+
+	Challenge []byte `json:"challenge"`
+	NumUnits  uint32 `json:"num_units"`
+}
+
+type CertifyRequest struct {
+	Proof    ProofToCertify         `json:"proof"`
+	Metadata ProofToCertifyMetadata `json:"metadata"`
+}
+
+type CertifyResponse struct {
+	Certificate []byte `json:"certificate"`
+	Signature   []byte `json:"signature"`
+	PubKey      []byte `json:"pub_key"`
+}
+
+type Certifier struct {
+	logger *zap.Logger
+	db     *localsql.Database
+	client certifierClient
+}
+
+func NewCertifier(
+	db *localsql.Database,
+	logger *zap.Logger,
+	client certifierClient,
+) *Certifier {
+	c := &Certifier{
+		client: client,
+		logger: logger,
+		db:     db,
+	}
+
+	return c
+}
+
+func (c *Certifier) GetCertificate(poet string) *certifier.PoetCert {
```

Comment:
NIT:

```suggestion
func (c *Certifier) Certificate(poet string) *certifier.PoetCert {
```

I'm also not sure about the usefulness of the method in general. There are two places where it is called:

1. in the `NIPostBuilder` where it's only purpose is to swallow the `sql.ErrNoFound` error - arguably its simpler if the `NIPostBuilder` fetches the certificate directly from the DB and (debug) log if it (didn't) find it or if an error occurred
2. the certify method below - where it can just be inlined

Reply:
Its purpose is also to implement the `certifierService` interface. The fact that the cert is stored in the local DB is an implementation detail of the `Certifier` that implements this interface.

Reply:
It seems convoluted from a users perspective. Ideally I would like to just pass whatever I want to submit to the poet client and the poet client handles the authentication.

Instead the nipost builder now has to 1. try to fetch a certificate and then pass that certificate to the poet client or 2. fail to fetch a certificate and then decide to generate one with `Recertify` or use PoW instead. Imo the poet client should do that for the user (either when instantiating it or when calling submit) - the nipost builder shouldn't care about communication details with poet.

Reply:
That would be best, but I wouldn't put it in the poet HTTP client, but create a _poet service_ thing containing all the poet-specific logic, including submitting to all known/registered Poets. The `nipostBuilder` would then only need to take the challenge and call `poetService.Submit(challenge)`. The logic inside `poetService` would figure out the details (PoW, Cert, registering in N poets, retries). But this is a bigger change and deserves a dedicated PR

Reply:
Sure, if we do not include it here we should create an issue for it and tackle it in the future 🙂

Reply:
Created issue #5865.